sunpy/io/special/genx.py

```diff
@@ -0,0 +1,140 @@
+import xdrlib
+from collections import OrderedDict
+from functools import partial
+import copy
+
+import numpy as np
+
+class ssw_xdrlib(xdrlib.Unpacker):
+    """
+    `xdrlib.Unpacker` customisation to read strings as written by IDL.
+    """
+    def unpack_string(self):
+        n = self.unpack_uint()
+        if n > 0:
+            n = self.unpack_uint()
+        return self.unpack_fstring(n).decode()
+
+def read_struct_skeleton(xdrdata):
+    """
+    Reads the skeleton of the IDL's structure as written in
+    solarsoft `build_str()` function.
+    """
+    # Read Number of tags
+    ntags = xdrdata.unpack_uint()
+    # Read names of tags
+    tags = [xdrdata.unpack_string() for n in range(ntags)]
+    # Read the tag type
+    tagdict = OrderedDict()
+    for tt in tags:
+        dim = xdrdata.unpack_uint()
+        arr_size = xdrdata.unpack_farray(dim + 2, xdrdata.unpack_int) # [7, 1]
+        typedata = arr_size[-2]
+        if typedata == 8: # it's a structure
+            if dim == 2 and arr_size[0] == 1:
+                # For when structures has been defined with 2 dim but only has one:
+                # bb = replicate({tata: 1, bebe:2, casa:'asa'}, 3)
+                # dd = replicate({tata: 1, bebe:2, casa:'asa'}, 1, 3)
+                # print,size(bb), size(dd)
+                #            1           3           8           3
+                #            2           1           3           8           3
+                dim = 1
+                arr_size[0] = arr_size[1]
+            if arr_size[-1] > 1:
+                tagdict[tt] = np.array([read_struct_skeleton(xdrdata)] * arr_size[-1]).reshape(arr_size[0:-2])
+            else:
+                tagdict[tt] = read_struct_skeleton(xdrdata)
+        else:
+            tagdict[tt] = [dim] + arr_size
+    return tagdict
+
+def unsuported_dtype(dtype):
+    raise ValueError("Can't read {} data type".format(dtype))
+
+def struct_to_data(xdrdata, subskeleton):
+    """"
+    Converts the dictionary with the keys and IDL's size output to
```

Comment:
This could be a little more descriptive, saying how the `subskeleton` needs to be the output of `read_struct_skeleton` and how it modified the values in place.